File: python_solver/rules.py
from __future__ import annotations

from datetime import datetime

from .types import (
    ActiveEntry,
    Defaults,
    EntityConfig,
    StateObject,
    ThresholdStep,
    WhenCondition,
)

_UNAVAILABLE_STATES = frozenset(("unavailable", "unknown"))


def evaluate_entity(
    config: EntityConfig,
    states: dict[str, StateObject],
    tiers: list[str],
    defaults: Defaults,
    now: datetime | None = None,
) -> ActiveEntry | None:
    if now is None:
        now = datetime.now()

    glyph_name = _resolve_glyph(config, states)
    state_obj = states.get(config.entity_id)
    state_str: str | None = state_obj.state if state_obj is not None else None
    is_unavailable = state_str is None or state_str in _UNAVAILABLE_STATES

    if is_unavailable and not _has_unavailable_override(config, state_str, states):
        if defaults.unavailable_action == "hide":
            return None
        fallback_tier = tiers[-1] if tiers else "status"
        return ActiveEntry(
            entity_config=config,
            tier=fallback_tier,
            color=None,
            glyph_name=glyph_name,
            show_info=defaults.show_info,
            focus_mode=False,
            indicator_only=False,
            drive_zone_indicator=False,
        )

    if config.rules:
        return _evaluate_rules(
            config, state_str or "", states, tiers, defaults, now, glyph_name
        )

    if config.thresholds:
        return _evaluate_thresholds(config, state_str or "", defaults, glyph_name)

    return None


def apply_focus_mode(
    entries: list[ActiveEntry], tiers: list[str]
) -> list[ActiveEntry]:
    # Caller must supply tiers in most-urgent-first order; tiers[0] is treated as the critical tier.
    if not any(e.focus_mode for e in entries):
        return list(entries)
    most_urgent = tiers[0] if tiers else ""
    return [e for e in entries if e.tier == most_urgent]


# ── private helpers ────────────────────────────────────────────────────────────


def _resolve_glyph(config: EntityConfig, states: dict[str, StateObject]) -> str:
    if config.glyph == "entity":
        obj = states.get(config.entity_id)
        return obj.attributes.get("icon", "") if obj is not None else ""
    return config.glyph or ""


def _has_unavailable_override(
    config: EntityConfig,
    state_str: str | None,
    states: dict[str, StateObject],
) -> bool:
    if not config.rules:
        return False
    check = {state_str} if state_str is not None else _UNAVAILABLE_STATES
    # When state is None (entity missing), any explicit unavailable/unknown rule is an override.
    # When state is a concrete string, only an exact-match rule is an override.
    # A rule with 'also' conditions is only a reliable override when those conditions currently hold.
    for r in config.rules:
        if r.when.state not in check:
            continue
        if r.when.also:
            if all(
                (obj := states.get(cond.entity_id)) is not None and obj.state == cond.state
                for cond in r.when.also
            ):
                return True
        else:
            return True
    return False


def _match_when(
    when: WhenCondition,
    state_str: str,
    states: dict[str, StateObject],
    now: datetime,
) -> bool:
    if when.state is not None and state_str != when.state:
        return False

    if when.range is not None:
        try:
            val = float(state_str)
        except ValueError:
            return False
        lo, hi = when.range
        if lo is not None and val < lo:
            return False
        if hi is not None and val > hi:
            return False

    if when.above is not None:
        try:
            val = float(state_str)
        except ValueError:
            return False
        if val <= when.above:
            return False

    if when.time_range is not None and not _check_time_range(when.time_range, now):
        return False

    if when.also:
        for cond in when.also:
            other = states.get(cond.entity_id)
            if other is None or other.state != cond.state:
                return False

    return True


def _check_time_range(time_range: tuple[str, str], now: datetime) -> bool:
    current = now.strftime("%H:%M")
    start, end = time_range
    if start <= end:
        return start <= current < end
    # Midnight-crossing window (e.g. "22:00" to "06:00")
    return current >= start or current < end


def _evaluate_rules(
    config: EntityConfig,
    state_str: str,
    states: dict[str, StateObject],
    tiers: list[str],
    defaults: Defaults,
    now: datetime,
    glyph_name: str,
) -> ActiveEntry | None:
    assert config.rules is not None
    for rule in config.rules:
        if not _match_when(rule.when, state_str, states, now):
            continue
        then = rule.then
        if then.action == "hide":
            return None
        tier = then.tier or (tiers[-1] if tiers else "status")
        color = then.color
        show_info = then.show_info if then.show_info is not None else defaults.show_info
        indicator_only = then.action == "indicator"
        # indicator_only: glyph hidden; then.indicator: glyph shown but zone still driven
        drive_zone = indicator_only or then.indicator
        return ActiveEntry(
            entity_config=config,
            tier=tier,
            color=color,
            glyph_name=glyph_name,
            show_info=show_info,
            focus_mode=then.focus_mode,
            indicator_only=indicator_only,
            drive_zone_indicator=drive_zone,
        )
    return None


def _evaluate_thresholds(
    config: EntityConfig,
    state_str: str,
    defaults: Defaults,
    glyph_name: str,
) -> ActiveEntry | None:
    assert config.thresholds is not None
    try:
        val = float(state_str)
    except ValueError:
        return None

    color_scale = config.color_scale or defaults.color_scale
    matched_step: ThresholdStep | None = None
    matched_index = -1

    # Thresholds must be in strictly ascending 'above' order (validated by validate_entity_config).
    # Iterating forward and keeping the last match gives the highest threshold the value has crossed.
    for i, step in enumerate(config.thresholds):
        if val > step.above:
            matched_step = step
            matched_index = i

    if matched_step is None:
        return None

    if matched_step.color is not None:
        color = matched_step.color
    elif color_scale:
        color = color_scale[matched_index % len(color_scale)]
    else:
        color = None

    return ActiveEntry(
        entity_config=config,
        tier=matched_step.tier,
        color=color,
        glyph_name=glyph_name,
        show_info=defaults.show_info,
        focus_mode=False,
        indicator_only=False,
        drive_zone_indicator=False,
    )

File: src/solver/rules.ts
import type {
  EntityConfig,
  ActiveEntry,
  Defaults,
  StateObject,
  Rule,
  ThresholdStep,
  WhenCondition,
} from './types';
import { entityDefaultGlyph } from '../utils/glyph';

function resolveEntityIconName(entityId: string, stateObj: StateObject | undefined): string {
  const attrIcon = stateObj?.attributes?.icon as string | undefined;
  if (attrIcon) return attrIcon;
  const deviceClass = stateObj?.attributes?.device_class as string | undefined;
  return entityDefaultGlyph(entityId, deviceClass);
}

// Parse "HH:MM" to minutes since midnight
function toMinutes(hhmm: string): number {
  const [h, m] = hhmm.split(':').map(Number);
  return h * 60 + m;
}

function inTimeRange(start: string, end: string, now: Date): boolean {
  const cur = now.getHours() * 60 + now.getMinutes();
  const s = toMinutes(start);
  const e = toMinutes(end);
  if (s <= e) {
    return cur >= s && cur <= e;
  }
  // midnight-crossing window
  return cur >= s || cur <= e;
}

function matchesWhen(
  when: WhenCondition,
  state: string,
  states: Record<string, StateObject>,
  now: Date,
): boolean {
  if (when.state !== undefined && state !== when.state) return false;

  if (when.range !== undefined) {
    const val = parseFloat(state);
    if (isNaN(val)) return false;
    const [low, high] = when.range;
    if (low !== null && val < low) return false;
    if (high !== null && val > high) return false;
  }

  if (when.above !== undefined) {
    const val = parseFloat(state);
    if (isNaN(val) || val <= when.above) return false;
  }

  if (when.time_range !== undefined) {
    const [start, end] = when.time_range;
    if (!inTimeRange(start, end, now)) return false;
  }

  if (when.also !== undefined) {
    for (const cond of when.also) {
      const s = states[cond.entity];
      if (!s || s.state !== cond.state) return false;
    }
  }

  return true;
}

function evaluateRules(
  config: EntityConfig,
  state: string,
  states: Record<string, StateObject>,
  tiers: string[],
  resolvedGlyph: string,
  defaults: Defaults,
  now: Date,
): ActiveEntry | null {
  if (!config.rules) return null;

  for (const rule of config.rules as Rule[]) {
    if (!matchesWhen(rule.when, state, states, now)) continue;

    const then = rule.then;

    if (then.action === 'hide') return null;

    return {
      entityConfig: config,
      tier: then.tier ?? tiers[0],
      color: then.color ?? '',
      glyphName: resolvedGlyph,
      showInfo: then.show_info ?? defaults.show_info ?? false,
      focusMode: then.focus_mode ?? false,
      indicatorOnly: then.action === 'indicator',
      driveZoneIndicator: then.action === 'show' && (then.indicator === true),
    };
  }

  return null;
}

function evaluateThresholds(
  config: EntityConfig,
  state: string,
  tiers: string[],
  resolvedGlyph: string,
  defaults: Defaults,
): ActiveEntry | null {
  if (!config.thresholds) return null;

  const val = parseFloat(state);
  if (isNaN(val)) return null;

  const colorScale = config.color_scale ?? defaults.color_scale ?? [];

  // Find highest threshold that val exceeds (value > above)
  let matched: ThresholdStep | null = null;
  let matchedIndex = -1;
  for (let i = 0; i < config.thresholds.length; i++) {
    const step = config.thresholds[i];
    if (val > step.above) {
      matched = step;
      matchedIndex = i;
    }
  }

  if (!matched) return null;

  const color = matched.color ?? (matchedIndex < colorScale.length ? colorScale[matchedIndex] : '');

  return {
    entityConfig: config,
    tier: matched.tier ?? tiers[0],
    color,
    glyphName: resolvedGlyph,
    showInfo: defaults.show_info ?? false,
    focusMode: false,
    indicatorOnly: false,
    driveZoneIndicator: false,
  };
}

export function evaluateEntity(
  config: EntityConfig,
  states: Record<string, StateObject>,
  tiers: string[],
  defaults: Defaults,
  now: Date = new Date(),
): ActiveEntry | null {
  if (tiers.length === 0) return null;

  const stateObj = states[config.entity_id];
  const rawState = stateObj?.state;
  const isUnavailable =
    !stateObj || rawState === 'unavailable' || rawState === 'unknown';

  if (isUnavailable) {
    const hasExplicitRule = config.rules?.some(
      (r) => r.when.state === 'unavailable' || r.when.state === 'unknown',
    ) ?? false;

    if (!hasExplicitRule) {
      if (!defaults.unavailable_action || defaults.unavailable_action === 'hide') {
        return null;
      }
      // unavailable_action === 'show': return minimal entry at least-urgent tier
      const resolvedGlyph =
        config.glyph === 'entity'
          ? resolveEntityIconName(config.entity_id, stateObj)
          : config.glyph ?? '';
      return {
        entityConfig: config,
        tier: tiers[tiers.length - 1],
        color: '',
        glyphName: resolvedGlyph,
        showInfo: defaults.show_info ?? false,
        focusMode: false,
        indicatorOnly: false,
        driveZoneIndicator: false,
      };
    }
  }

  const state = rawState ?? '';

  // Pre-resolve "entity" glyph
  const resolvedGlyph =
    config.glyph === 'entity'
      ? resolveEntityIconName(config.entity_id, stateObj)
      : config.glyph ?? '';

  if (config.rules) {
    return evaluateRules(config, state, states, tiers, resolvedGlyph, defaults, now);
  }

  if (config.thresholds) {
    return evaluateThresholds(config, state, tiers, resolvedGlyph, defaults);
  }

  return null;
}

export function applyFocusMode(
  entries: ActiveEntry[],
  tiers: string[],
): ActiveEntry[] {
  if (!entries.some((e) => e.focusMode)) return entries;
  return entries.filter((e) => e.tier === tiers[0]);
}
